clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java

```diff
@@ -2412,15 +2412,36 @@ protected long measureLagWithCallToPubSub(
       PubSubTopic topic,
       int partition,
       long currentOffset) {
-    if (currentOffset < 0) {
+    return measureLagWithCallToPubSub(pubSubServerName, topic, partition, currentOffset, this::getTopicManager);
+  }
+
+  protected static long measureLagWithCallToPubSub(
+      String pubSubServerName,
+      PubSubTopic topic,
+      int partition,
+      long currentOffset,
+      Function<String, TopicManager> topicManagerProvider) {
+    if (currentOffset < OffsetRecord.LOWEST_OFFSET) {
+      // -1 is a valid offset, which means that nothing was consumed yet, but anything below that is invalid.
       return Long.MAX_VALUE;
     }
-    TopicManager tm = getTopicManager(pubSubServerName);
-    long endOffset = tm.getLatestOffsetCached(topic, partition) - 1;
+    TopicManager tm = topicManagerProvider.apply(pubSubServerName);
+    long endOffset = tm.getLatestOffsetCached(topic, partition);
     if (endOffset < 0) {
+      // A negative value means there was a problem in measuring the end offset, and therefore we return "infinite lag"
       return Long.MAX_VALUE;
+    } else if (endOffset == 0) {
+      // The topic is empty and therefore, by definition, there cannot be any lag
+      return 0;
     }
-    return endOffset - currentOffset;
+
+    /**
+     * An empty topic should have an end offset of zero. A topic with a single message in it will have an end offset of
```

Comment:
So I guess it would be more accurate to say that a topic with end offset of zero has always been empty. Whereas a topic which had messages in it before may now be empty (due either to time retention or to log compaction where all keys have gotten a null value [which we do not do, but anyway...]). In such case where a topic has had messages before but is empty now, the end offset would be > 0.